File: October 22/18 Oct/38. Count and Say LC.cpp
class Solution {
public:
    string countAndSay(int n) {
        if(n==1) return "1";
        else if(n==2) return "11";
        string prev="1",curr="";
        int i=0;
        char prev_ch='$';
        int len,prev_count=0;
        while(--n)
        {
            len=prev.size();
            i=0;
            prev_ch=prev[0];
            prev_count=1;
            for(int i=1;i<len;i++)
            {
                if(prev[i]==prev_ch) prev_count++;
                else
                {
                    curr+=to_string(prev_count);
                    curr+=prev_ch;
                    prev_ch=prev[i];
                    prev_count=1;
                }
            }
            curr+=to_string(prev_count);
            curr+=prev_ch;
            prev=curr;
            curr="";
        }
        
        return prev;
    }
};

File: APRIL 2025/18 April/38. Count and Say LC.java
class Solution {
    class DigitCount {
        int digit;
        int count;
        public DigitCount(int digit, int count)
        {
            this.digit = digit;
            this.count = count;
        }
    }
    public String countAndSay(int n) {
        String str = "1";
        for(int i=0; i<(n-1); i++)
        {
            List<DigitCount> count = countDigits(str);
            str = makeString(count);
        }
        return str;
    }
    List<DigitCount> countDigits(String str)
    {
        List<DigitCount> count = new ArrayList<>();
        int N = str.length(), freq=0;
        char last = '$';
        for(int i=0; i<N; i++)
        {
            if(str.charAt(i)==last)
                freq++;
            else
            {
                if(last!='$')
                    count.add(new DigitCount((int)(last-'0'), freq));
                last = str.charAt(i);
                freq = 1;
            }
            // System.out.print(str.charAt(i)+" ");
        }
        count.add(new DigitCount((int)(last-'0'), freq));
        // System.out.println();
        return count;
    }
    String makeString(List<DigitCount> count)
    {
        String str = "";
        for(DigitCount dc:count)
        {
            str = str+Integer.toString(dc.count);
            str = str+Integer.toString(dc.digit);
            
        }
        return str;
    }
}
